orbs/git/orb.yml

```diff
@@ -0,0 +1,72 @@
+---
+version: 2.1
+
+description: |
+  Alternate checkout command to that provided by CircleCI
+
+  CircleCI's `checkout` command has two major issues [[1][1],[2][2]]:
+  - it does shallow clones: you get the latest state of
+    your repository, but don't really get any history;
+  - it messes with branch references;
+
+  Those are not an issue, until your CI workflow has reasons
+  to switch branches or look at the code's history.
+
+  If you do any of those things, you can use this alternate
+  `custom_checkout` command.
+
+  Of course, if CircleCI does things the way it does, is for
+  performance reasons. With time this command may become
+  slower than the original one, on especially large
+  repositories.
+
+  [1]: https://discuss.circleci.com/t/the-checkout-step-mangles-branches-messes-the-history/24975
+  [2]: https://discuss.circleci.com/t/git-checkout-of-a-branch-destroys-local-reference-to-master/23781
+
+commands:
+  custom_checkout:
+    steps:
+    - run:
+        name: "Custom checkout"
+        command: |
+          set -e
+
+          if [ "${HOME}" = "/" ]
+          then
+            export HOME=$(getent passwd $(id -un) | cut -d: -f6)
+          fi
+
+          mkdir -p ~/.ssh
+
+          # yamllint disable-line rule:line-length
+          echo 'github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ==' >> ~/.ssh/known_hosts
+          # yamllint disable-line rule:line-length
+          echo 'bitbucket.org ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==' >> ~/.ssh/known_hosts
```

Comment:
**question**

Is it ok to hardcode those values?

I recently saw a similar use case in a github action and I think the recommendation was to use `ssh-keyscan`.

A bit like this:
```
ssh-keyscan github.com >> ~/.ssh/known_hosts
ssh-keyscan bitbucket.org >> ~/.ssh/known_hosts
```

Reply:
This is the way CircleCI does it. This is mostly a modified version of their `checkout` commands.
Security-wise there is no issue those are public keys.

If github or bitbucket changed there public keys we'll know it fast i guess. And yes we'd have to update those values. But they don't change that often. And it is better to have to change them once here than let people pass parameters or whatever and have to change many configurations files.

Whether to use `ssh-keyscan` depends on is it installed in all docker images ? I doubt so.